src/validation.cpp

```diff
@@ -2359,6 +2360,87 @@ CScript calculate_contract(const CScript& federationRedeemScript, const CScript&
     return scriptDestination;
 }
 
+bool GetAmountFromParentChainPegin(CAmount& amount, const Sidechain::Bitcoin::CTransaction& txBTC, unsigned int nOut)
+{
+    amount = txBTC.vout[nOut].nValue;
+    return true;
+}
+
+bool GetAmountFromParentChainPegin(CAmount& amount, const CTransaction& txBTC, unsigned int nOut)
+{
+    if (!txBTC.vout[nOut].nValue.IsExplicit()) {
+        return false;
+    }
+    amount = txBTC.vout[nOut].nValue.GetAmount();
+    return true;
+}
+
+template<typename T>
+static bool GetBlockAndTxFromMerkleBlock(uint256& block_hash, uint256& tx_hash, T& merkle_block, const std::vector<unsigned char>& merkle_block_raw)
+{
+    try {
+        std::vector<uint256> tx_hashes;
+        std::vector<unsigned int> tx_indices;
+        CDataStream merkle_block_stream(merkle_block_raw, SER_NETWORK, PROTOCOL_VERSION);
+        merkle_block_stream >> merkle_block;
+        block_hash = merkle_block.header.GetHash();
+
+        if (!merkle_block_stream.empty()) {
+           return false;
+        }
+        if (merkle_block.txn.ExtractMatches(tx_hashes, tx_indices) != merkle_block.header.hashMerkleRoot || tx_hashes.size() != 1) {
+            return false;
+        }
+        tx_hash = tx_hashes[0];
+    } catch (std::exception& e) {
+        // Invalid encoding of merkle block
+        return false;
+    }
+    return true;
+}
+
+template<typename T>
+static bool CheckPeginTx(const std::vector<unsigned char>& tx_data, T& pegtx, const COutPoint& prevout, const CAmount claim_amount, const CScript& claim_script)
+{
+    try {
+        CDataStream pegtx_stream(tx_data, SER_NETWORK, PROTOCOL_VERSION);
+        pegtx_stream >> pegtx;
```

Comment:
The community seems to be okay with this (paraphrased) construct but it just seems wrong to me:

```
    shared_ptr<const CTransaction> pegtx;  //Uninitialized shared_ptr!!
    pegtx_stream >> pegtx;  //Looks like a bug - how are we to know pegtx is initialized before written to?
```

What be the downside of just using *Unserialize* instead?
```
    shared_ptr<const CTransaction> pegtx;
    Unserialize(pegtx_stream, pegtx);
```

Reply:
It's doing that under the hood, so I don't see how it could be an uninitialized shared pointer.

Reply:
Yes, I get that the initialization is happening under the hood in Unserialize.  It's not really a bug but it looked like it could be one to me (as a noob in this code base anyway) while doing a casual reading of the code. It's NBD, it just made the code a little harder to grok for me.

Reply:
I just wanted to leave the code as similar as it was before as possible, but I can use 
```
shared_ptr<const CTransaction> pegtx;
    Unserialize(pegtx_stream, pegtx);
```
if that's preferred.

Reply:
as-is is fine imo.